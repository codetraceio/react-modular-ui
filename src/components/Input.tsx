import * as React from 'react';

import {
  default as AbstractComponent,
  IModifiers
} from './AbstractComponent';

export interface IInputProps {
  size?: string | number;
  view?: string;
  color?: string;
  disabled?: boolean;
  name?: string;
  label?: string;
  value?: string;
  placeholder?: string;
  shape?: string;

  onChange?: (value: string, event: React.ChangeEvent<HTMLInputElement>) => void;
  onKeyDown?: (value: string, event: React.KeyboardEvent<HTMLInputElement>) => void;
  onKeyUp?: (value: string, event: React.KeyboardEvent<HTMLInputElement>) => void;
  onKeyPress?: (value: string, event: React.KeyboardEvent<HTMLInputElement>) => void;
  onFocus?: (value: string, event: React.FocusEvent<HTMLInputElement>) => void;
  onBlur?: (value: string, event: React.FocusEvent<HTMLInputElement>) => void;
  onSubmit?: (value: string, event: React.KeyboardEvent<HTMLInputElement>) => void;
  onClick?: (value: string, event: React.MouseEvent<HTMLInputElement>) => void;
}

export default class Input extends AbstractComponent<IInputProps, {}> {
  getModifierObject(): IModifiers {
    return {
      size: this.props.size,
      view: this.props.view,
      color: this.props.color,
      disabled: this.props.disabled,
      shape: this.props.shape,
    };
  }

  onEvent(
    event: React.SyntheticEvent<HTMLInputElement>,
    callback: (value: string, event: React.SyntheticEvent<HTMLInputElement>) => void
  ) {
    if (typeof callback === 'function') {
      callback(event.currentTarget.value, event);
    }
  }

  onKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
    if (typeof this.props.onKeyDown === 'function') {
      this.props.onKeyDown(event.currentTarget.value, event);
    }
    if (event.isPropagationStopped()) {
      return;
    }
    if (
      typeof this.props.onSubmit === 'function' &&
      ['Enter', 'NumpadEnter'].indexOf(event.key) !== -1 && !event.shiftKey
    ) {
      this.props.onSubmit(event.currentTarget.value, event);
    }
  }

  renderLabel() {
    return this.props.label ? (
      <div className={this.getElementName('input', 'label')}>{this.props.label}</div>
    ) : null;
  }

  render() {
    return (
      <div className={this.getBlockName('input', this.getModifierObject())}>
        {this.renderLabel()}
        <input
          name={this.props.name}
          placeholder={this.props.placeholder}
          value={this.props.value}
          disabled={this.props.disabled}
          onChange={(event: React.ChangeEvent<HTMLInputElement>) => this.onEvent(event, this.props.onChange)}
          onKeyDown={(event: React.KeyboardEvent<HTMLInputElement>) => this.onKeyDown(event)}
          onKeyUp={(event: React.KeyboardEvent<HTMLInputElement>) => this.onEvent(event, this.props.onKeyUp)}
          onKeyPress={(event: React.KeyboardEvent<HTMLInputElement>) => this.onEvent(event, this.props.onKeyPress)}
          onFocus={(event: React.FocusEvent<HTMLInputElement>) => this.onEvent(event, this.props.onFocus)}
          onBlur={(event: React.FocusEvent<HTMLInputElement>) => this.onEvent(event, this.props.onBlur)}
          onClick={(event: React.MouseEvent<HTMLInputElement>) => this.onEvent(event, this.props.onClick)}
        />
      </div>
    )
  }
}
