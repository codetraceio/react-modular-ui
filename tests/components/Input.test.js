import * as React from 'react';
import * as renderer from 'react-test-renderer';

import Input from '../../src/components/Input';

describe('Input', () => {
  it('should render Input', () => {
    const tree = renderer.create(
      <Input
        size="4"
        color="primary"
        view="flat"
        disabled
        onClick={() => {}}
        onBlur={() => {}}
        onChange={() => {}}
        onFocus={() => {}}
        onKeyDown={() => {}}
        onKeyUp={() => {}}
        onKeyPress={() => {}}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
