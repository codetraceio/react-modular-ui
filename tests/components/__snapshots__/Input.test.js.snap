// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render Input 1`] = `
<div
  className="c-input -size-4 -view-flat -color-primary -disabled"
>
  <input
    disabled={true}
    name={undefined}
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    placeholder={undefined}
    value={undefined}
  />
</div>
`;
